### DISCUSSION TITLE
command to align  text from right to left (rtl)

### DISCUSSION THREAD

**Discussion by noor-tg** [helix-editor/helix#10311] [General]
hi, 
first thank you for making this amazing code editor.

I want a command to allow to align all the document text from right to left (rtl). all the document or per line.
this can be used for languages like Arabic, Urdu. so if I have markdown document contains this languages letters . I can use this command to show it from right to left as expected.

this is how markdown with arabic language show in helix:

![helix-markdown-arabic](https://user-images.githubusercontent.com/60521739/184479133-e6f8596d-13af-40aa-8869-35347a7eb97a.png)

thank you.

**Comment by A-Walrus** [comment]
This is quite tricky for multiple reasons. One of them is that terminal emulators may exhibit different behaviors when dealing with RTL text. 
For example, `Kitty` by default, will flip the order of RTL characters within a word, but not the order of words, while other terminal emulators don't flip anything at all. 
Kitty:
![image](https://user-images.githubusercontent.com/58790821/184861306-287cf493-b9e0-41f6-b124-300467693976.png)
Alacritty:
![image](https://user-images.githubusercontent.com/58790821/184861357-a318810b-f66f-4be8-adad-72f8410edb96.png)

This might have to wait until there is a GUI version of Helix, where we can have full control over text rendering.

**Comment by noor-tg** [comment]
I use kitty with hilex . and I toke the screenshot above when using it.

so. there is no way to handle text from right to left (words and characters) in terminal ?

**Comment by A-Walrus** [comment]
Pretty much. There have been some attempts, for example with Vim, but in my opinion none of them work well enough to be generally useful.
Bidirectional text is hard, even more so in a terminal.

**Comment by noor-tg** [comment]
ok , thank you @A-Walrus .

**Comment by zefr0x** [comment]
> This is quite tricky for multiple reasons. One of them is that terminal emulators may exhibit different behaviors when dealing with RTL text. For example, `Kitty` by default, will flip the order of RTL characters within a word, but not the order of words, while other terminal emulators don't flip anything at all. Kitty: ![image](https://user-images.githubusercontent.com/58790821/184861306-287cf493-b9e0-41f6-b124-300467693976.png) Alacritty: ![image](https://user-images.githubusercontent.com/58790821/184861357-a318810b-f66f-4be8-adad-72f8410edb96.png)
> 
> This might have to wait until there is a GUI version of Helix, where we can have full control over text rendering.

RTL/Arabic support in Kitty and Alacritty is just broken (it is not about different behaviors). No one should try to fix it from the applications' side.

If any one would work in it the new [cosmic-term](https://github.com/pop-os/cosmic-term) has the greatest support i've ever seen in a terminal for RTL/Arabic and bidirectional text.
KDE's Konsole is also great, GNOME Terminal is good, but some small things are still broken on them regarding character rendering and spacing between characters.

I believe that if Arabic will be supported in Helix inside a terminal, it should be on top of a terminal that supports it e.g. [cosmic-term](https://github.com/pop-os/cosmic-term), no other terminal with broken support should be considered.

For now on top of [cosmic-term](https://github.com/pop-os/cosmic-term) there is no problem in Helix that I can see other than text alignment to the left (it should be to the right).

**Comment by menisadi** [comment]
Just noting that at least on kitty you can turn off the flipping behavior. It's called `force_ltr` on the conf file.
Also, RTL on Vim (and Neovim) works ok in my opinions, but not bidirectional text which I guess is much harder to implement. There are tools like [Fribidi](https://github.com/fribidi/fribidi) which try to bring bidirectional text to the terminal but embedding it inside an editor does sound like a headache. 
I heard that Emacs has a better solution, but from what I understand it is possible due to Emacs being sort of an OS of its own (as opposed to Vim and Helix.
